File: app/src/main/java/com/letianpai/robot/desktop/apps/adapter/AppListVerticalViewPagerAdapter.java
package com.letianpai.robot.desktop.apps.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.PagerAdapter;

import java.util.ArrayList;
import java.util.List;

/**
 * 纵向应用列表Adapter
 * @author liujunbin
 */
public class AppListVerticalViewPagerAdapter extends PagerAdapter {
    private List<View> appListViews = new ArrayList<>();
    private Context mContext;


    public AppListVerticalViewPagerAdapter(List<View> appList, Context context){
        this.appListViews.addAll(appList);
        this.mContext = context;
    }
    @Override
    public int getCount() {
        return appListViews.size();
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object o) {
        return view == o;
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        if (position < appListViews.size()){
            View view = appListViews.get(position);
            container.addView(view);
            return view;
        }
        return null;
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        if ((appListViews.size()>0)&&(position < appListViews.size())){
            container.removeView(appListViews.get(position));
        }
    }


    public void refresh(List<View> appList) {
        this.appListViews.clear();
        this.appListViews.addAll(appList);
//        for (int i = 0; i < appListViews.size(); i++){
//            this.appListViews.set(i,appList.get(i));
//        }
        notifyDataSetChanged();
    }


    @Override
    public int getItemPosition(Object object) {
        return POSITION_NONE;
    }


}

File: app/src/main/java/com/letianpai/robot/desktop/apps/adapter/AppListVerticalViewPagerAdapter.kt
package com.letianpai.robot.desktop.apps.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter

/**
 * Vertical Application List Adapter
 * @author liujunbin
 */
class AppListVerticalViewPagerAdapter(appList: List<View>, context: Context?) :
    PagerAdapter() {
    private val appListViews: MutableList<View> = ArrayList()
    private val mContext: Context?


    init {
        appListViews.addAll(appList)
        this.mContext = context
    }

    override fun getCount(): Int {
        return appListViews.size
    }

    override fun isViewFromObject(view: View, o: Any): Boolean {
        return view === o
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        return if (position < appListViews.size) {
            val view = appListViews[position]
            container.addView(view)
            view
        } else {
            val placeholderView = View(container.context)
            container.addView(placeholderView)
            placeholderView
        }
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        if ((appListViews.size > 0) && (position < appListViews.size)) {
            container.removeView(appListViews[position])
        }
    }


    fun refresh(appList: List<View>) {
        appListViews.clear()
        appListViews.addAll(appList)
        //        for (int i = 0; i < appListViews.size(); i++){
//            this.appListViews.set(i,appList.get(i));
//        }
        notifyDataSetChanged()
    }


    override fun getItemPosition(`object`: Any): Int {
        return POSITION_NONE
    }
}
